doc/man/ulf_gen.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.13.
.TH ULF_GEN "1" "June 2022" "ulf_gen (link2next@gmail.com)," "User Commands"
.SH NAME
ulf_gen v1.6.0, written by Stanley Hong \- ulf_gen
.SH SYNOPSIS
.B ulf_gen
\fI\,-L <ulc-file> \/\fR[\fI\,options\/\fR] \fI\,<file|dir>\/\fR
.SH DESCRIPTION
.IP
ulf_gen generates the frequencies of the keywords of a language.
.IP
ulf_gen \fB\-L\fR <ulc\-file> <file1> <file2> ...
ulf_gen \fB\-L\fR <ulc\-file> \fB\-o\fR \fI\,/usr/local/share/a.ulf\/\fP <file1> <file2> ...
ulf_gen \fB\-L\fR <ulc\-file> \fB\-l\fR <list\-file> <target\-dir>
.TP
\fB\-L\fR, \fB\-\-lang=\fR<ulc\-spec>
Specify the lexical\-spec(*.ulc) of the language
.TP
\fB\-l\fR, \fB\-\-list=\fR<list\-file>
Specify the list of data source files
.TP
\fB\-o\fR, \fB\-\-output=\fR<a\-file>
Specify the the output filepath(*.ulf)
.TP
\fB\-v\fR, \fB\-\-verbose\fR
verbose mode
.TP
\fB\-V\fR, \fB\-\-version\fR
Print the version information
.TP
\fB\-h\fR, \fB\-\-help\fR
Display the short help
.PP
ulf_gen collects the usage statistics of given language by sampling from source code files.
It dumps the frequencies of keywords.
After saving them in the file suffixed by '*.ulf', you can use it together with its ulc\-file.
For example, Let the output file of 'sample.ulc' be 'sample.ulf'.
The file 'sample.ulf' is together read if only it exists in the same directory of sample.ulc.
The ulf\-file is optional but can accelerates the generated lexical tokenizer.
.PP
To get the ulf\-file of 'sample.ulc' from listed source code files,
.IP
ulf_gen \fB\-L\fR sample.ulc a.c b.c src/d.c
.PP
To save the output\-file to other file path than default, use o\-option.
.IP
ulf_gen \fB\-o\fR \fI\,/opt/share/b.ulf\/\fP \fB\-L\fR sample src/main.c src/main.h
.PP
To process all the file paths in the list file 'a.list'
.IP
ulf_gen \fB\-L\fR sample \fB\-l\fR a.list \fI\,/package/home\/\fP
.PP
This processes all the file paths joined by '/package/home'.
.IP
is provided under The MIT License (MIT).
.SH "SEE ALSO"
The full documentation for
.B ulf_gen
is maintained as a Texinfo manual.  If the
.B info
and
.B ulf_gen
programs are properly installed at your site, the command
.IP
.B info ulf_gen
.PP
should give you access to the complete manual.


CHANGES

  v1.8.15
  -----------
    o bug-fixes on numbers
    o Refactoring token numbers
      The format of the API function uls_lexeme(), which is the current token string, is changed
         and the uls_tokstr() is now a little different from uls_lexeme().
      a) Integers: The supported radices of number are 2, 3, 4, 8, 16, and 10.
         The corresponding prefixes 0b, 0t, 0q, 0o, 0x are prepended in the token string, uls_lexeme().
         uls_tokstr() will gives you it as hexa-decimal format prefixed by '0x'.
      b) Floating-Point numbers:
         uls_lexeme() will represent it as the one similar to scientific E-notation and
         uls_tokstr() gives you it as the common form.
      c) Use uls_number_suffix() to get the number suffix.
      For more details, see the user-guide 'uls_ug-0.8.2.pdf' in the folder 'doc'.

  v1.8.14
  -----------
    o Restored & Reinforced the incomplete wide string system of uls.
       You can build the examples in wide string mode
      a) On Windows select the charset to take effect the constant '_UNICODE'
      b) On Linux define constant ULS_USE_WSTR to build programs in 'examples'.
           Refer to 'Makefile.am'
    o Fixed the bug that cause abnormal termination at the end of program
          when you use both the libraries libuls and libulscompat.
    o bug-fixes on
      a) generating c files from uld files with new token name 'NULL'.
      b) printing error messages in the wide string mode

  v1.8.13
  -----------
    o The support of Java is included in the main source tree.
    o The user guide(uls_ug-0.8.1.pdf) has been updated

  v1.8.12
  -----------
    o bug-fixes on
      a) the generated cpp source files from uld-file
    o Delete the debris of ANSI-encoding api, which is not supported any more.
    o Fixed the example of uls_push_isrc in tests/dump_toks
    o Moved ulc2yaml from 'tests' to 'examples' directory.

  v1.8.11
  -----------
    o bug-fixes on
      a) data loss when returning nested input
      b) reading utf8 file containing multibytes chars.
      c) reading utf32 file
      d) reading utf files with BOM(8,16,32)

  v1.8.10
  -----------
    o bug-fixes on
      a) the length of the wide string of lexeme
      b) detecting unterminated string when keyword is defined in ulc-file
      c) the procedure of getting escaped string
    o Removed duplicated resetting the struct xcontext
    o Can safely set the param 'lexeme' of uls_set_tok() to NULL
    o Added the token ELLIPSIS in ulc_exam/ansi_ulc
    o VisualStudio 2022 is needed to build the solution

  v1.8.9
  -----------
    o bug-fixes on
      a) processing token EOI
      b) Not yet support generating source file other than c/c++ from uld-file
      c) line-numbering when reading ulc-files

  v1.8.8
  -----------
    o bug-fixes on
      a) generating cpp source file from ulc-file
      b) using ulf-file, together with ulc-file
      c) adding uld alias from uld-file

  v1.8.7
  -----------
    o bug-fixes on
      a) 1-char punctuation token which is explicitly defined in ulc-file
      b) Unicode punctuation token of length more than 2.
      c) uls_get_extra_tokdef()

  v1.8.6
  -----------
    o Added a program 'tools/ulc2yaml' to dump yaml-file from ulc-file.
    o a few bug fixes

  v1.8.5
  -----------
    o Changed the 'character set' properties of Visual Studio projects
       from 'Multibytes' to 'Unicode'.
    o Use utf-8 string instead of ANSI-string.
        Add the option '/utf-8' in properties of projects to build your own cpp-applications

  v1.8.4
  -----------
    o The feature of pre-processing input files by converting CR to LF is deleted
    o Refactoring the structure of arrays shared by C & C#
    o a few bug fixes related to unicode encoding, line-numbering of tokens.

 v1.8.3
  -----------
    o Fixed the installation process on Windows 10.
    o The programs & libraries on Windows is successfully built and tested
           in the following environment:
        a) Windows 10
        b) Visual Studio 2019
             Solutioin Configuration : Release
             Solution Platform : x64, Win32
        c) Platform Toolset : v142
        d) Windows SDK Version : 10.0.x
        e) Target Framework: .NET Framework 4.8

  v1.8.2
  -----------
    o The secondary search path for the sysprops file(uls_sysprops.txt) is added.
        1) C:\Program Files\Common Files\UlsWin
        2) C:\opt\shared\UlsWin
    o To run the examples of C#.NET, C++.NET,
        it's necessary to adjust the path of the UlsDotNet.dll.
        The path of it is set to 'C:\opt\UlsWin\Bin\UlsDotNet.dll' in each project.
        You can modify it properly at the 'Reference' of the projects.
    o The table for character group should be cleared at loading ulc-file on Windows.
    o The missing document file is added.

  v1.8.1
  -----------
    o The Windows version of uls source code is pulished with name of 'uls-windows'.
    o The examples of C++.NET are included in the folder 'CppNetExamples'.


  v1.8
  -----------
    o The format of ulc-file is upgraded.
      a) Decimal separator is supported for number token.
         The separator may be one character.
         For example, you can specify the separator in ulc-file.
            DECIMAL_SEPARATOR: _
      b) To specify the radix of number, you can use attribute NUMBER_PREFIXES in ulc-file.
            NUMBER_PREFIXES: 0:8 0x16 0b:2 0o:8
         This is the list of pairs where the item is number prefix and radix.
      c) The format of the attribute 'QUOTE_TYPE:' of ulc-file is largely changed.

      The upgrade_ulc.py in the directory 'tools' will convert the previous format to v2.3.
      You can use it as following examples:
         i) upgrade_ulc.py -o new_sample.ulc sample.ulc
            This will convert input file 'sample.ulc' to output file 'new_sample.ulc'.
        ii) upgrade_ulc.py sample1.ulc sample2.ulc
            This will convert two file 'sample1.ulc' and 'sample2.ulc'.

tools/ulf_gen/ulf_gen.c
/*
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

/*
  <file> ulf_gen.c </file>
  <brief>
    Generating ULF(Unified Lexical Frequencies) file gathering statistics of keywords usage.
    This file is part of ULS, Unified Lexical Scheme.
  </brief>
  <author>
    Stanley Hong <link2next@gmail.com>, August 2011.
  </author>
*/

#define ULS_DECL_PROTECTED_PROC
#include "uls.h"
#include "uls/uls_misc.h"
#include "uls/uls_freq.h"

#include "ult_utils.h"
#include <ctype.h>

#define THIS_PROGNAME "ulf_gen"
#define DFL_N_SAMPLES 0xFFFF

_ULS_DEFINE_STRUCT(stat_of_round)
{
	uls_dflhash_state_t dflhash_stat0;
	int   n; // # of buckets
	double avg, sigma2;
	double gamma;
};

char *progname;
char home_dir[ULS_FILEPATH_MAX+1];
const char *config_file;
char *target_dir;
char *filelist;
FILE *fp_list;
char *out_file;
char out_file_buff[ULS_FILEPATH_MAX+1];
int  opt_verbose;

uls_lex_ptr_t sam_lex;
uls_hashfunc_t ulf_hashfunc;

int target_hashtable_size;
int n_samples;

#define ULFGEN_OPTSTR "L:l:o:n:s:vVHh"

#ifdef HAVE_GETOPT
#include <getopt.h>

static const struct option longopts[] = {
	{ "lang",  required_argument,      NULL, 'L' },
	{ "list",  required_argument,      NULL, 'l' },
	{ "output",  required_argument,    NULL, 'o' },
	{ "num-iter",  required_argument,  NULL, 'n' },
	{ "size-hash",  required_argument, NULL, 's' },
	{ "verbose",  no_argument,         NULL, 'v' },
	{ "version", no_argument,          NULL, 'V' },
	{ "Help",    no_argument,          NULL, 'H' },
	{ "help",    no_argument,          NULL, 'h' },
	{ NULL, 0, NULL, 0 }
};
#endif

static void usage_synopsis(void)
{
	uls_printf("Usage: %s -L <ulc-file> [options] <file|dir>\n", progname);
	uls_printf("  %s generates the frequencies of the keywords of a language.\n", progname);
	uls_printf("       %s -L <ulc-file> <file1> <file2> ...\n", progname);
	uls_printf("       %s -L <ulc-file> -o /usr/local/share/a.ulf <file1> <file2> ...\n", progname);
	uls_printf("       %s -L <ulc-file> -l <list-file> <target-dir>\n", progname);
}

static void usage_desc(void)
{
#ifdef ULS_WINDOWS
	uls_printf("  -L <ulc-spec>      Specify the lexical-spec(*.ulc) of the language\n");
	uls_printf("  -l <list-file>     Specify the list of data source files\n");
	uls_printf("  -o <a-file>        Specify the the output filepath(*.ulf)\n");
	uls_printf("  -v, --verbose      verbose mode\n");
	uls_printf("  -V, --version      Print the version information\n");
	uls_printf("  -h, --help         Display the short help\n");
#else
	uls_printf("  -L, --lang=<ulc-spec>   Specify the lexical-spec(*.ulc) of the language\n");
	uls_printf("  -l, --list=<list-file>  Specify the list of data source files\n");
	uls_printf("  -o, --output=<a-file>   Specify the the output filepath(*.ulf)\n");
	uls_printf("  -v, --verbose           verbose mode\n");
	uls_printf("  -V, --version           Print the version information\n");
	uls_printf("  -h, --help              Display the short help\n");
#endif
}

static void usage_brief(void)
{
	usage_synopsis();
	uls_printf("\n");

	usage_desc();
	uls_printf("\n");
}

static void usage(void)
{
	usage_brief();
}

static void usage_long(void)
{
	usage_brief();

	uls_printf("%s collects the usage statistics of given language", progname);
	uls_printf(" by sampling from source code files.\n");
	uls_printf("It dumps the frequencies of keywords.\n");
	uls_printf("After saving them in the file suffixed by '*.ulf', you can use it together with its ulc-file.\n");
	uls_printf("For example, Let the output file of 'sample.ulc' be 'sample.ulf'.\n");
	uls_printf("The file 'sample.ulf' is together read if only it exists in the same directory of sample.ulc.\n");
	uls_printf("The ulf-file is optional but can accelerates the generated lexical tokenizer.\n");
	uls_printf("\n");

	uls_printf("To get the ulf-file of 'sample.ulc' from listed source code files,\n");
	uls_printf("  %s -L sample.ulc a.c b.c src/d.c\n", progname);
	uls_printf("\n");

	uls_printf("To save the output-file to other file path than default, use o-option.\n");
	uls_printf("  %s -o /opt/share/b.ulf -L sample src/main.c src/main.h\n", progname);
	uls_printf("\n");

	uls_printf("To process all the file paths in the list file 'a.list'\n");
	uls_printf("  %s -L sample -l a.list /package/home\n", progname);
	uls_printf("This processes all the file paths joined by '/package/home'.\n");
	uls_printf("\n");
}

static int ulfgen_options(int opt, char* optarg)
{
	int   stat = 0;
	int   siz;

	switch (opt) {
	case 'L':
		config_file = optarg;
		break;

	case 'l':
		filelist = ult_strdup(optarg);
		uls_path_normalize(filelist, filelist);
		break;

	case 'n':
		n_samples = atoi(optarg);
		if (n_samples <= 0) n_samples = 1;
		break;

	case 'o':
		out_file = ult_strdup(optarg);
		uls_path_normalize(out_file, out_file);
		break;

	case 's':
		if ((siz=atoi(optarg)) > 0) {
			target_hashtable_size =  siz;
		} else {
			err_log("non-positive hash-size(%d) specified!", siz);
			stat = -1;
		}
		break;

	case 'v':
		++opt_verbose;
		break;

	case 'H':
		usage_long();
		stat = 3;
		break;

	case 'h':
		usage();
		stat = 2;
		break;

	case 'V':
		uls_printf("%s %s, written by %s,\n\tis provided under %s.\n",
			progname, ULF_GEN_PROGVER, ULS_AUTHOR, ULS_LICENSE_NAME);
		stat = 1;
		break;

	default:
		err_log("undefined option -%c", opt);
		stat = -1;
		break;
	}

	return stat;
}

static int
parse_options(int argc, char* argv[])
{
#ifdef HAVE_GETOPT
	int   rc, opt, longindex;
#endif
	int   i0;

	if (ult_getcwd(home_dir, sizeof(home_dir)) < 0)
		err_panic("%s: fail to getcwd()", __FUNCTION__);

	target_hashtable_size = ULF_HASH_TABLE_SIZE;

	filelist = NULL;
	out_file = NULL;
	target_dir = NULL;
	n_samples = DFL_N_SAMPLES;

#ifdef HAVE_GETOPT
	while ((opt=getopt_long(argc, argv, ULFGEN_OPTSTR, longopts, &longindex)) != -1) {
		if ((rc=ulfgen_options(opt, optarg)) != 0) {
			if (rc > 0) rc = 0;
			return rc;
		}
	}
	i0 = optind;
#else
	if ((i0=uls_getopts(argc, argv, ULFGEN_OPTSTR, ulfgen_options)) <= 0) {
		return i0;
	}
#endif

	if (config_file == NULL) {
		err_log("specify the path of config-file(*.ulc)");
		return -1;
	}

	if (filelist != NULL) {
		if (i0 >= argc) {
			err_log("Specify the target-directory to apply the paths in %s\n", filelist);
			return -1;
		}

		if ((fp_list=uls_fp_open(filelist, ULS_FIO_READ)) == NULL) {
			err_log("%s: fail to read '%s'", __FUNCTION__, filelist);
			return -1;
		}

		target_dir = ult_strdup(argv[i0++]);
		uls_path_normalize(target_dir, target_dir);
	}

	return i0;
}

int
incl_if_keyw(uls_keyw_stat_list_t *ks_lst, const char* keyw)
{
	uls_keyw_stat_ptr_t kwstat;
	int stat = 0;

	kwstat = ulc_search_kwstat_list(ks_lst, keyw);

	if (kwstat != uls_nil) {
		if (kwstat->freq < INT_MAX) {
			++kwstat->freq;
			stat = 1;
		} else {
			stat = -1;
		}
	}

	return stat;
}

int
proc_file(char *fpath, uls_keyw_stat_list_t *ks_lst)
{
	if (fpath == NULL) return 0;

	if (uls_push_file(sam_lex, fpath, 0) < 0) {
		return -1;
	}

	if (opt_verbose >= 1)
		err_log("processing %s, ...", fpath);

	for ( ; ; ) {
		uls_get_tok(sam_lex);
		if (uls_is_eoi(sam_lex)) break;

		incl_if_keyw(ks_lst, uls_lexeme(sam_lex));
	}

	return 1;
}

int
proc_filelist(FILE *fin, uls_keyw_stat_list_t *ks_lst)
{
	char  linebuff[ULS_FILEPATH_MAX+1], *lptr, *fpath;
	int   i, len, lno = 0, stat=0;

	while (1) {
		if ((len=uls_fp_gets(fin, linebuff, sizeof(linebuff), 0)) <= ULS_EOF) {
			if (len < ULS_EOF) {
				err_log("%s: io-error", __FUNCTION__);
				stat = -1;
			}
			break;
		}

		++lno;

		lptr = skip_blanks(linebuff);
		if (*lptr=='\0' || *lptr=='#') continue;

		fpath = lptr;
		for (i=(int)strlen(fpath)-1; i>=0; i--) {
			if (!isspace(fpath[i])) break;
		}
		fpath[++i] = '\0';

		if (proc_file(fpath, ks_lst) < 0) {
			err_log("fail to process '%s'", fpath);
			stat = -1;
			break;
		}
	}

	return stat;
}

static void
uls_hashfunc_set_params(stat_of_round_ptr_t p_stat,
	int n_slots, int n_shifts, uls_uint32 hcodes0)
{
	p_stat->dflhash_stat0.n_slots = n_slots;
	p_stat->dflhash_stat0.n_shifts = n_shifts;
	p_stat->dflhash_stat0.init_hcode = hcodes0;

	p_stat->n = 0;
	p_stat->gamma = 0.;
}

void
go_round(uls_keyw_stat_list_t *ks_lst,
	uls_uint32 hcodes0, int* hash_range, int n_hash_range, stat_of_round_ptr_t p_stat)
{
	uls_ref_parray(lst, keyw_stat);
	uls_decl_parray_slots(slots_lst, keyw_stat);
	uls_keyw_stat_ptr_t e;
	int i, hash_id, n;
	double sum, sum1, sum2;

	memset(hash_range, 0x00, n_hash_range * sizeof(int));

	uls_hashfunc_set_params(p_stat, n_hash_range, -1, hcodes0);

	sum = 0.;
	lst = uls_ptr(ks_lst->lst);
	slots_lst = uls_parray_slots(lst);
	for (i=0; i < ks_lst->lst.n; i++) {
		e = slots_lst[i];

		hash_id = ulf_hashfunc(&p_stat->dflhash_stat0, e->keyw);
		if (hash_range[hash_id] < INT_MAX)
			++hash_range[hash_id];

		if (opt_verbose >= 3)
			uls_printf(" '%s' --> %d\n", e->keyw, hash_id);

		sum += hash_range[hash_id] * e->freq;
	}

	sum1 = sum2 = 0.;
	n = 0;
	for (hash_id=0; hash_id < n_hash_range; hash_id++) {
		if ((i=hash_range[hash_id]) > 0) {
			sum1 += i;
			sum2 += i*i;
			++n;
		}
	}

	p_stat->avg = sum1 / n;
	p_stat->sigma2 = sum2 / n - p_stat->avg * p_stat->avg;
	p_stat->n = n;
	p_stat->gamma = sum / n;
}

void
get_random_hcodes(int n_hcodes0, uls_uint32* hcodes0)
{
	uls_uint32 hcode;
	int i, j;

	for (i=0; i<n_hcodes0; i++) {
		hcode = 0;
		for (j=0; j<4; j++) {
			hcode ^= rand();
			hcode <<= 8;
		}
		hcodes0[i] = hcode;
	}
}

void
calc_good_hcode0(uls_keyw_stat_list_t *ks_lst, stat_of_round_ptr_t p_stat)
{
	int *bucket_size_array;
	stat_of_round_t  round_stat;
	int     i;

	bucket_size_array = (int *) uls_malloc(target_hashtable_size * sizeof(int));

	for (i=0; i<n_samples; i++) {
		get_random_hcodes(1, &p_stat->dflhash_stat0.init_hcode);

		go_round(ks_lst, p_stat->dflhash_stat0.init_hcode,
			bucket_size_array, target_hashtable_size, uls_ptr(round_stat));

		if (opt_verbose >= 2 && i % 1000 == 0) {
			uls_printf("\tn=%d: avg = %.3f, sigma2 = %.3f, gamma=%.3f\n",
				round_stat.n, round_stat.avg, round_stat.sigma2, round_stat.gamma);
		}

		if (i > 0) {
			if (round_stat.gamma < p_stat->gamma) {
				*p_stat = round_stat;
			}
		} else {
			*p_stat = round_stat;
		}
	}

	if (opt_verbose >= 1) {
		uls_printf("\tavg = %.3f, sigma2 = %.3f, n=%d, gamma=%.3f\n",
			p_stat->avg, p_stat->sigma2, p_stat->n, p_stat->gamma);
	}

	uls_mfree(bucket_size_array);

	if (opt_verbose >= 2) {
		uls_printf(" hcode[] = ");
		uls_printf(" 0x%08X", p_stat->dflhash_stat0.init_hcode);
		uls_printf("\n");
	}
}

static int keyw_stat_comp_by_freq_reverse(const uls_voidptr_t a, const uls_voidptr_t b)
{
	uls_keyw_stat_ptr_t a1 = (uls_keyw_stat_ptr_t) a;
	uls_keyw_stat_ptr_t b1 = (uls_keyw_stat_ptr_t) b;

	return b1->freq - a1->freq;
}

int
ulf_create_file_internal(FILE *fp_list, FILE *fp_out, int i0, int n_args, char **args)
{
	uls_keyw_stat_list_t *keyw_stat_list;
	stat_of_round_t best_round_stat;
	int i;

	keyw_stat_list = ulc_export_kwtable(uls_ptr(sam_lex->idkeyw_table));
	if (keyw_stat_list == uls_nil) {
		err_log("No keyword information!");
		return -1;
	}

	if (fp_list != NULL) {
		if (ult_chdir(target_dir) < 0) {
			err_log("can't change to %s", target_dir);
			ulc_free_kwstat_list(keyw_stat_list);
			return -1;
		}

		if (proc_filelist(fp_list, keyw_stat_list) < 0) {
			err_log("Failed to process files in %s", target_dir);
			return -1;
		}

		if (ult_chdir(home_dir) < 0) {
			err_log("fail to chdir(%s)", home_dir);
			ulc_free_kwstat_list(keyw_stat_list);
			return -1;
		}

	} else {
		for (i=i0; i<n_args; i++) {
			proc_file(args[i], keyw_stat_list);
		}
	}

	_uls_quicksort_vptr(uls_parray_slots(uls_ptr(keyw_stat_list->lst)), keyw_stat_list->lst.n, keyw_stat_comp_by_freq_reverse);

	calc_good_hcode0(keyw_stat_list, uls_ptr(best_round_stat));

	ulf_create_file(1, &best_round_stat.dflhash_stat0.init_hcode,
		target_hashtable_size, keyw_stat_list, fp_out);

	ulc_free_kwstat_list(keyw_stat_list);

	return 0;
}

int
main(int argc, char* argv[])
{
	FILE *fp_out;
	const char *conf_fname;
	int i0, i, conf_fname_len, stat = 0;
	int cse_insen;

	initialize_uls();

	progname = THIS_PROGNAME;
	if (argc <= 1) {
		usage_brief();
		return 1;
	}

	if (ulc_prepend_searchpath_pwd() < 0) {
		err_log("InternalError: don't know about the program '%s'.", THIS_PROGNAME);
		return -1;
	}

	if ((i0=parse_options(argc, argv)) <= 0) {
		if (i0 < 0) err_log("Incorrect use of command options.");
		return i0;
	}

	conf_fname = uls_filename(config_file, &conf_fname_len);

	if (out_file == NULL) {
		strcpy(out_file_buff, conf_fname);
		i = conf_fname_len;
		strcpy(out_file_buff+i, ".ulf");
		out_file = out_file_buff;
	}

	if (opt_verbose)
		ulc_list_searchpath(config_file);

	if ((sam_lex=uls_create(config_file)) == uls_nil) {
		err_log("can't create a uls-object for %s.", config_file);
		return -1;
	}

	cse_insen = sam_lex->flags & ULS_FL_CASE_INSENSITIVE;
	ulf_hashfunc = uls_get_hashfunc(ULS_HASH_ALGORITHM, cse_insen);

	if ((fp_out=uls_fp_open(out_file, ULS_FIO_CREAT)) == NULL) {
		err_log("%s: fail to create '%s'", __FUNCTION__, out_file);
		stat = -1;
	} else {
		err_log("Gathering the statistics of keywords usage, ...");
		if (ulf_create_file_internal(fp_list, fp_out, i0, argc, argv) < 0) {
			err_log("%s: internal error!'", __FUNCTION__);
			stat = -1;
		} else {
			err_log("Writing the frequencies of keywords to %s, ...", out_file);
		}
		uls_fp_close(fp_out);
	}

	if (fp_list != NULL) {
		uls_fp_close(fp_list);
		fp_list = NULL;
	}

	uls_destroy(sam_lex);

	uls_mfree(filelist);
	uls_mfree(out_file);
	uls_mfree(target_dir);

	return stat;
}
